In [2]:
import pandas as pd
import glob
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display
import missingno as msno
import plotly.express as px
from scipy.stats import chi2_contingency, mannwhitneyu, kstest, ttest_ind
from plotly.offline import plot
from scipy.stats import zscore


In [3]:
Data = pd.read_csv("Data.csv")
Data.head()

<ipython-input-3-2ab91b33fe0a>:1: DtypeWarning:

Columns (0,10,11,21,22,23,29,37,52,53,56,58,69) have mixed types. Specify dtype option on import or set low_memory=False.



,Partición,CONSECUTIVE,COD_EVE,FEC_NOT,SEMANA,ANO,COD_PRE,COD_SUB,EDAD,UNI_MED,...,Pais_ocurrencia,Nombre_evento,Departamento_ocurrencia,Municipio_ocurrencia,Pais_residencia,Departamento_residencia,Municipio_residencia,Departamento_Notificacion,Municipio_notificacion,COD_EVE.1
0,Datos_2014_875,2934558,875,2014-06-03,22,2014,7600107759,2,15,1,...,COLOMBIA,VCM VIF VSX,VALLE,CALI,NaN,VALLE,CALI,VALLE,CALI,NaN
1,Datos_2014_875,2934568,875,2014-03-24,13,2014,5200101457,18,17,1,...,COLOMBIA,VCM VIF VSX,NARIÑO,PASTO,NaN,NARIÑO,PASTO,NARIÑO,PASTO,NaN
2,Datos_2014_875,2934589,875,2014-08-12,33,2014,1100110146,0,17,1,...,COLOMBIA,VCM VIF VSX,BOGOTA,BOGOTA,NaN,BOGOTA,BOGOTA,BOGOTA,BOGOTA,NaN
3,Datos_2014_875,2934609,875,2014-05-18,21,2014,6800103534,1,17,1,...,COLOMBIA,VCM VIF VSX,SANTANDER,PIEDECUESTA,NaN,SANTANDER,PIEDECUESTA,SANTANDER,BUCARAMANGA,NaN
4,Datos_2014_875,2934620,875,2014-03-26,5,2014,6819000713,2,17,1,...,COLOMBIA,VCM VIF VSX,SANTANDER,LANDAZURI,NaN,SANTANDER,LANDAZURI,SANTANDER,CIMITARRA,NaN


In [4]:
print(f"Cantidad de registros: {Data.shape[0]}, Cantidad de variables: {Data.shape[1]}")

Cantidad de registros: 1014375, Cantidad de variables: 75


In [5]:
DeleteColumns = [
    'Partición', 'COD_PRE', 'COD_EVE', 'COD_SUB', 'COD_ASE', 'AJUSTE', 'CER_DEF', 'Nombre_evento',
    'FEC_ARC_XL', 'FEC_AJU', 'FM_FUERZA', 'FM_UNIDAD', 'FM_GRADO',
    'confirmados', 'consecutive_origen', 'va_sispro', 'COD_EVE.1'
]
Data = Data.drop(columns=DeleteColumns)

In [6]:
print(f"Filas: {Data.shape[0]}, Columnas: {Data.shape[1]}")

Filas: 1014375, Columnas: 58


In [7]:
# Conversión a cadenas (string)
Data['CONSECUTIVE'] = Data['CONSECUTIVE'].astype(str)
Data['UNI_MED'] = Data['UNI_MED'].astype(str)
Data['nacionalidad'] = Data['nacionalidad'].astype(str)
Data['nombre_nacionalidad'] = Data['nombre_nacionalidad'].astype(str)
Data['SEXO'] = Data['SEXO'].astype(str)
Data['COD_PAIS_O'] = Data['COD_PAIS_O'].astype(str)
Data['COD_DPTO_O'] = Data['COD_DPTO_O'].astype(str)
Data['COD_MUN_O'] = Data['COD_MUN_O'].astype(str)
Data['AREA'] = Data['AREA'].astype(str)
Data['OCUPACION'] = Data['OCUPACION'].astype(str)
Data['TIP_SS'] = Data['TIP_SS'].astype(str)
Data['PER_ETN'] = Data['PER_ETN'].astype(str)
Data['GRU_POB'] = Data['GRU_POB'].astype(str)
Data['nom_grupo'] = Data['nom_grupo'].astype(str)
Data['GP_DISCAPA'] = Data['GP_DISCAPA'].astype(str)
Data['GP_DESPLAZ'] = Data['GP_DESPLAZ'].astype(str)
Data['GP_MIGRANT'] = Data['GP_MIGRANT'].astype(str)
Data['GP_CARCELA'] = Data['GP_CARCELA'].astype(str)
Data['GP_GESTAN'] = Data['GP_GESTAN'].astype(str)
Data['GP_INDIGEN'] = Data['GP_INDIGEN'].astype(str)
Data['GP_POBICFB'] = Data['GP_POBICFB'].astype(str)
Data['GP_MAD_COM'] = Data['GP_MAD_COM'].astype(str)
Data['GP_DESMOVI'] = Data['GP_DESMOVI'].astype(str)
Data['GP_PSIQUIA'] = Data['GP_PSIQUIA'].astype(str)
Data['GP_VIC_VIO'] = Data['GP_VIC_VIO'].astype(str)
Data['GP_OTROS'] = Data['GP_OTROS'].astype(str)
Data['fuente'] = Data['fuente'].astype(str)
Data['COD_PAIS_R'] = Data['COD_PAIS_R'].astype(str)
Data['COD_DPTO_R'] = Data['COD_DPTO_R'].astype(str)
Data['COD_MUN_R'] = Data['COD_MUN_R'].astype(str)
Data['COD_DPTO_N'] = Data['COD_DPTO_N'].astype(str)
Data['COD_MUN_N'] = Data['COD_MUN_N'].astype(str)
Data['TIP_CAS'] = Data['TIP_CAS'].astype(str)
Data['PAC_HOS'] = Data['PAC_HOS'].astype(str)
Data['CON_FIN'] = Data['CON_FIN'].astype(str)
Data['CBMTE'] = Data['CBMTE'].astype(str)
Data['Estado_final_de_caso'] = Data['Estado_final_de_caso'].astype(str)
Data['nom_est_f_caso'] = Data['nom_est_f_caso'].astype(str)
Data['Nom_upgd'] = Data['Nom_upgd'].astype(str)
Data['Pais_ocurrencia'] = Data['Pais_ocurrencia'].astype(str)
Data['Departamento_ocurrencia'] = Data['Departamento_ocurrencia'].astype(str)
Data['Municipio_ocurrencia'] = Data['Municipio_ocurrencia'].astype(str)
Data['Pais_residencia'] = Data['Pais_residencia'].astype(str)
Data['Departamento_residencia'] = Data['Departamento_residencia'].astype(str)
Data['Municipio_residencia'] = Data['Municipio_residencia'].astype(str)
Data['Departamento_Notificacion'] = Data['Departamento_Notificacion'].astype(str)
Data['Municipio_notificacion'] = Data['Municipio_notificacion'].astype(str)

# Conversión a fechas
Data['FEC_NOT'] = pd.to_datetime(Data['FEC_NOT'], format='%Y-%m-%d', errors='coerce')
Data['FEC_CON'] = pd.to_datetime(Data['FEC_CON'], format='%Y-%m-%d', errors='coerce')
Data['INI_SIN'] = pd.to_datetime(Data['INI_SIN'], format='%Y-%m-%d', errors='coerce')
Data['FEC_HOS'] = pd.to_datetime(Data['FEC_HOS'], format='%Y-%m-%d', errors='coerce')
Data['FEC_DEF'] = pd.to_datetime(Data['FEC_DEF'], format='%Y-%m-%d', errors='coerce')

# Conversión a numéricos (enteros)
Data['SEMANA'] = pd.to_numeric(Data['SEMANA'], errors='coerce').astype('Int64')
Data['ANO'] = pd.to_numeric(Data['ANO'], errors='coerce').astype('Int64')
Data['EDAD'] = pd.to_numeric(Data['EDAD'], errors='coerce').astype('Int64')
Data['sem_ges'] = pd.to_numeric(Data['sem_ges'], errors='coerce').astype('Int64')

## **Análisis descriptivo inicial**

Tras la asignación correcta de los tipos de datos, se realizó un análisis descriptivo inicial utilizando la función ´data.describe()´. Este análisis proporcionó una visión general de las variables numéricas y de fecha, permitiendo identificar tendencias, rangos y valores atípicos.

In [8]:
display(Data.describe())

,FEC_NOT,SEMANA,ANO,EDAD,sem_ges,FEC_CON,INI_SIN,FEC_HOS,FEC_DEF
count,1014375,1014375.0,1014375.0,1014375.0,25313.0,1014354,940478,155896,1419
mean,2018-06-10 22:44:42.030609664,27.001046,2017.914265,22.834042,19.614269,2018-06-06 13:03:02.386031872,2018-05-24 21:27:43.434125824,2019-01-24 10:06:41.436855296,2017-12-28 17:24:13.699788032
min,2012-01-01 00:00:00,1.0,2012.0,0.0,1.0,2007-01-01 00:00:00,1993-01-30 00:00:00,2011-11-07 00:00:00,2012-01-24 00:00:00
25%,2016-03-29 00:00:00,14.0,2016.0,11.0,10.0,2016-03-25 00:00:00,2016-03-03 00:00:00,2017-01-09 18:00:00,2015-10-06 00:00:00
50%,2018-10-05 00:00:00,27.0,2018.0,19.0,19.0,2018-10-01 00:00:00,2018-10-13 00:00:00,2019-06-27 00:00:00,2017-12-17 00:00:00
75%,2021-01-22 00:00:00,40.0,2021.0,32.0,28.0,2021-01-19 00:00:00,2021-02-17 00:00:00,2021-07-10 00:00:00,2020-07-05 00:00:00
max,2022-12-31 00:00:00,53.0,2022.0,131.0,45.0,2022-12-31 00:00:00,2022-12-31 00:00:00,2022-12-31 00:00:00,2022-12-30 00:00:00
std,NaN,14.836164,3.01495,16.901137,10.987746,NaN,NaN,NaN,NaN


In [9]:
# Identificar columnas categóricas
CategoricalColumns = Data.select_dtypes(include=['object', 'category']).columns

# Verificar valores faltantes en columnas categóricas
MissingCategorical = Data[CategoricalColumns].isnull().sum()
print(MissingCategorical[MissingCategorical > 0])

estrato      426004
FECHA_NTO     22835
dtype: int64


In [10]:
# Convertir cadenas vacías y otros valores a NA
Data[CategoricalColumns] = Data[CategoricalColumns].replace(['', ' ', 'NaN', 'NULL', 'nan', 'null'], pd.NA)

# Verificar nuevamente los valores faltantes
MissingCategoricalUpdated = Data[CategoricalColumns].isnull().sum()
print(MissingCategoricalUpdated[MissingCategoricalUpdated > 0])

nacionalidad             426004
nombre_nacionalidad      426004
GRU_POB                  951130
nom_grupo                426004
estrato                  426004
fuente                   426004
COD_PAIS_R               654172
FECHA_NTO                 22835
CBMTE                   1013090
Pais_residencia          654172
Municipio_residencia          5
dtype: int64


In [11]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
CBMTE,99.873321
FEC_DEF,99.860111
sem_ges,97.504572
GRU_POB,93.765126
FEC_HOS,84.631325
COD_PAIS_R,64.490154
Pais_residencia,64.490154
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697


In [12]:
import plotly.express as px

# Creación del gráfico de barras interactivo
fig = px.bar(
    MissingPercentage,
    x=MissingPercentage.index,
    y=MissingPercentage.values,
    labels={'x': 'Variables', 'y': 'Porcentaje de Valores Faltantes'},
    title='Porcentaje de Valores Faltantes por Variable',
    color=MissingPercentage.values,
    color_continuous_scale='Blues',
    text=MissingPercentage.values.round(2)
)

# Personalización del gráfico
fig.update_layout(
    xaxis_title='Variables',
    yaxis_title='Porcentaje de Valores Faltantes',
    title_font_size=24,
    xaxis_tickangle=-45,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(
        family="Georgia",
        size=14,
        color="black"
    ),
    title_font=dict(
        family="Georgia",
        size=24,
        color="black"
    ),
    coloraxis_colorbar=dict(
        title='Porcentaje',
        titlefont=dict(family="Georgia", size=14),
        tickfont=dict(family="Georgia", size=12)
    )
)

fig.update_traces(textposition='outside')

# Mostrar el gráfico
fig.show()

Sustento teórico:
Imputación de valores faltantes en datos de salud:

En el análisis de datos de salud, es común imputar valores faltantes basándose en reglas lógicas o clínicas. Por ejemplo, si un paciente está vivo, no tiene sentido tener una fecha de defunción o una causa de muerte. Esto se alinea con las buenas prácticas de manejo de datos en salud, donde la imputación debe ser clínicamente válida (Little & Rubin, 2019).

Fuente: Little, R. J., & Rubin, D. B. (2019). Statistical Analysis with Missing Data. Wiley.

Manejo de datos faltantes en registros de mortalidad:

En los registros de mortalidad, es crucial asegurar que los datos faltantes se manejen de manera que no distorsionen los análisis. Si un individuo está vivo, los campos relacionados con la muerte deben ser marcados como no aplicables o imputados con un valor que refleje esta condición (World Health Organization, 2020).

Fuente: World Health Organization. (2020). International Guidelines for Certification and Classification (Coding) of COVID-19 as Cause of Death.

In [13]:
# Verificar valores faltantes en CBMTE por grupo CON_FIN
print("Valores faltantes en CBMTE por grupo CON_FIN:")
print(Data.groupby('CON_FIN')['CBMTE'].apply(lambda x: x.isnull().sum()))

# Verificar valores faltantes en FEC_DEF por grupo CON_FIN
print("Valores faltantes en FEC_DEF por grupo CON_FIN:")
print(Data.groupby('CON_FIN')['FEC_DEF'].apply(lambda x: x.isnull().sum()))

Valores faltantes en CBMTE por grupo CON_FIN:
CON_FIN
0        421
1    1012353
2        316
Name: CBMTE, dtype: int64
Valores faltantes en FEC_DEF por grupo CON_FIN:
CON_FIN
0        423
1    1012356
2        177
Name: FEC_DEF, dtype: int64


In [14]:
# Verificar valores únicos en CON_FIN
print("Valores únicos en CON_FIN:")
print(Data['CON_FIN'].unique())

Valores únicos en CON_FIN:
['1' '0' '2']


In [15]:
# Convertir CON_FIN a tipo numérico
Data['CON_FIN'] = pd.to_numeric(Data['CON_FIN'], errors='coerce')

# Verificar valores únicos en CON_FIN después de la conversión
print("Valores únicos en CON_FIN después de la conversión:")
print(Data['CON_FIN'].unique())

Valores únicos en CON_FIN después de la conversión:
[1 0 2]


In [16]:
# Verificar valores faltantes en CON_FIN
print("Valores faltantes en CON_FIN después de la conversión:")
print(Data['CON_FIN'].isnull().sum())

Valores faltantes en CON_FIN después de la conversión:
0


In [17]:
# Imputar "0000" en CBMTE para pacientes vivos (CON_FIN == 1)
Data.loc[Data['CON_FIN'] == 1, 'CBMTE'] = Data.loc[Data['CON_FIN'] == 1, 'CBMTE'].fillna("0000")

In [18]:
# Convertir FEC_DEF a datetime si no lo está ya
Data['FEC_DEF'] = pd.to_datetime(Data['FEC_DEF'], errors='coerce')

# Imputar "9999-12-31" en FEC_DEF para pacientes vivos (CON_FIN == 1)
fecha_no_fallecio = pd.to_datetime('2090-12-31')
Data.loc[Data['CON_FIN'] == 1, 'FEC_DEF'] = Data.loc[Data['CON_FIN'] == 1, 'FEC_DEF'].fillna(fecha_no_fallecio)

In [19]:
# Verificar valores faltantes en CBMTE por grupo CON_FIN
print("Valores faltantes en CBMTE por grupo CON_FIN después de la imputación:")
print(Data.groupby('CON_FIN')['CBMTE'].apply(lambda x: x.isnull().sum()))

# Verificar valores faltantes en FEC_DEF por grupo CON_FIN
print("Valores faltantes en FEC_DEF por grupo CON_FIN después de la imputación:")
print(Data.groupby('CON_FIN')['FEC_DEF'].apply(lambda x: x.isnull().sum()))

Valores faltantes en CBMTE por grupo CON_FIN después de la imputación:
CON_FIN
0    421
1      0
2    316
Name: CBMTE, dtype: int64
Valores faltantes en FEC_DEF por grupo CON_FIN después de la imputación:
CON_FIN
0    423
1      0
2    177
Name: FEC_DEF, dtype: int64


In [20]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
sem_ges,97.504572
GRU_POB,93.765126
FEC_HOS,84.631325
COD_PAIS_R,64.490154
Pais_residencia,64.490154
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697
estrato,41.996697
fuente,41.996697


In [21]:
import plotly.express as px

# Creación del gráfico de barras interactivo
fig = px.bar(
    MissingPercentage,
    x=MissingPercentage.index,
    y=MissingPercentage.values,
    labels={'x': 'Variables', 'y': 'Porcentaje de Valores Faltantes'},
    title='Porcentaje de Valores Faltantes por Variable',
    color=MissingPercentage.values,
    color_continuous_scale='Blues',
    text=MissingPercentage.values.round(2)
)

# Personalización del gráfico
fig.update_layout(
    xaxis_title='Variables',
    yaxis_title='Porcentaje de Valores Faltantes',
    title_font_size=24,
    xaxis_tickangle=-45,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(
        family="Georgia",
        size=14,
        color="black"
    ),
    title_font=dict(
        family="Georgia",
        size=24,
        color="black"
    ),
    coloraxis_colorbar=dict(
        title='Porcentaje',
        titlefont=dict(family="Georgia", size=14),
        tickfont=dict(family="Georgia", size=12)
    )
)

fig.update_traces(textposition='outside')

# Mostrar el gráfico
fig.show()

In [22]:
# Verificar valores únicos en GRU_POB
print("Valores únicos en GRU_POB:")
print(Data['GRU_POB'].unique())

Valores únicos en GRU_POB:
[<NA> '5.0' '16.0' '7.0' '9.0' '13.0' '5' '9' '16' '7' '13' '14' '14.0'
 '  ']


Análisis de la variable GRU_POB
Porcentaje de datos faltantes:

Tienes un 93.77% de datos faltantes en GRU_POB, lo cual es extremadamente alto. En la literatura, se considera que variables con más del 50-60% de datos faltantes suelen ser candidatas a eliminación, ya que no aportan suficiente información para el análisis (Allison, 2001; Schafer & Graham, 2002).

Relación con otras columnas:

GRU_POB está relacionada con otras columnas como GP_GESTAN (gestante), GP_MIGRANTE (migrante), etc., que indican si la persona pertenece o no a un grupo poblacional específico (1 o 2).

Estas columnas (GP_GESTAN, GP_MIGRANTE, etc.) tienen una cobertura de datos mucho mejor, ya que no tienen un porcentaje tan alto de valores faltantes.

Valores únicos en GRU_POB:

Los valores únicos en GRU_POB son: [<NA>, '5.0', '16.0', '7.0', '9.0', '13.0', '5', '9', '16', '7', '13', '14', '14.0', ' '].

Hay inconsistencias en los valores (por ejemplo, '5.0' y '5' representan lo mismo, pero están en diferentes formatos). Además, hay cadenas vacías (' ') que deben ser tratadas como valores faltantes.

Sustento teórico para eliminar GRU_POB
Alto porcentaje de datos faltantes:

Según Allison (2001), cuando una variable tiene más del 50-60% de datos faltantes, su utilidad en el análisis se reduce significativamente, y es preferible eliminarla para evitar sesgos o problemas en el modelo.

Fuente: Allison, P. D. (2001). Missing Data. Sage Publications.

Redundancia con otras variables:

Las columnas relacionadas (GP_GESTAN, GP_MIGRANTE, etc.) ya capturan la información sobre los grupos poblacionales de manera más detallada y con menos datos faltantes.

Mantener GRU_POB no agregaría información adicional útil, y su alto porcentaje de datos faltantes podría introducir ruido en el análisis.

Inconsistencias en los datos:

Los valores únicos en GRU_POB muestran inconsistencias (por ejemplo, '5.0' y '5'), lo que sugiere problemas de calidad en los datos. Corregir estos problemas no sería práctico dado el alto porcentaje de datos faltantes.

In [23]:
# Eliminar la columna GRU_POB
Data = Data.drop(columns=['GRU_POB'])

# Verificar que la columna ha sido eliminada
print("Columnas después de eliminar GRU_POB:")
print(Data.columns)

Columnas después de eliminar GRU_POB:
Index(['CONSECUTIVE', 'FEC_NOT', 'SEMANA', 'ANO', 'EDAD', 'UNI_MED',
       'nacionalidad', 'nombre_nacionalidad', 'SEXO', 'COD_PAIS_O',
       'COD_DPTO_O', 'COD_MUN_O', 'AREA', 'OCUPACION', 'TIP_SS', 'PER_ETN',
       'nom_grupo', 'estrato', 'GP_DISCAPA', 'GP_DESPLAZ', 'GP_MIGRANT',
       'GP_CARCELA', 'GP_GESTAN', 'sem_ges', 'GP_INDIGEN', 'GP_POBICFB',
       'GP_MAD_COM', 'GP_DESMOVI', 'GP_PSIQUIA', 'GP_VIC_VIO', 'GP_OTROS',
       'fuente', 'COD_PAIS_R', 'COD_DPTO_R', 'COD_MUN_R', 'COD_DPTO_N',
       'COD_MUN_N', 'FEC_CON', 'INI_SIN', 'TIP_CAS', 'PAC_HOS', 'FEC_HOS',
       'CON_FIN', 'FEC_DEF', 'FECHA_NTO', 'CBMTE', 'Estado_final_de_caso',
       'nom_est_f_caso', 'Nom_upgd', 'Pais_ocurrencia',
       'Departamento_ocurrencia', 'Municipio_ocurrencia', 'Pais_residencia',
       'Departamento_residencia', 'Municipio_residencia',
       'Departamento_Notificacion', 'Municipio_notificacion'],
      dtype='object')


In [24]:
print(f"Filas: {Data.shape[0]}, Columnas: {Data.shape[1]}")

Filas: 1014375, Columnas: 57


In [25]:
# Verificar valores faltantes en columnas relacionadas
columnas_relacionadas = ['GP_DISCAPA', 'GP_DESPLAZ', 'GP_MIGRANT', 'GP_CARCELA',
                        'GP_GESTAN', 'GP_INDIGEN', 'GP_POBICFB', 'GP_MAD_COM',
                         'GP_DESMOVI', 'GP_PSIQUIA', 'GP_VIC_VIO', 'GP_OTROS']
print("Valores faltantes en columnas relacionadas:")
print(Data[columnas_relacionadas].isnull().sum())

Valores faltantes en columnas relacionadas:
GP_DISCAPA    0
GP_DESPLAZ    0
GP_MIGRANT    0
GP_CARCELA    0
GP_GESTAN     0
GP_INDIGEN    0
GP_POBICFB    0
GP_MAD_COM    0
GP_DESMOVI    0
GP_PSIQUIA    0
GP_VIC_VIO    0
GP_OTROS      0
dtype: int64


In [26]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
sem_ges,97.504572
FEC_HOS,84.631325
COD_PAIS_R,64.490154
Pais_residencia,64.490154
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697
estrato,41.996697
fuente,41.996697
INI_SIN,7.284978


In [27]:
# Verificar el tipo de dato de sem_ges
print("Tipo de dato de sem_ges:")
print(Data['sem_ges'].dtype)

Tipo de dato de sem_ges:
Int64


In [28]:
# Imputar 0 en sem_ges para no gestantes (GP_GESTAN == 2)
Data.loc[Data['GP_GESTAN'] == 2, 'sem_ges'] = Data.loc[Data['GP_GESTAN'] == 2, 'sem_ges'].fillna(0)

In [29]:
# Verificar valores faltantes en sem_ges después de la imputación
print("Valores faltantes en sem_ges después de la imputación:")
print(Data['sem_ges'].isnull().sum())

# Verificar valores únicos en sem_ges
print("Valores únicos en sem_ges después de la imputación:")
print(Data['sem_ges'].unique())

Valores faltantes en sem_ges después de la imputación:
989062
Valores únicos en sem_ges después de la imputación:
<IntegerArray>
[<NA>,    3,   26,    7,   19,   32,   36,   30,    9,   34,   22,   14,   23,
    6,   31,   21,   27,   24,   18,    2,   25,   10,   12,   11,    5,   16,
   15,   29,   38,   35,   28,   17,    8,   20,   37,   39,   40,    1,   13,
    4,   33,   44,   45,   41,   42,   43]
Length: 46, dtype: Int64


In [30]:
# Verificar gestantes con sem_ges faltante después de la imputación
gestantes_con_sem_ges_faltante = Data[(Data['GP_GESTAN'] == 1) & (Data['sem_ges'].isna())]
print("Gestantes con sem_ges faltante después de la imputación:")
print(gestantes_con_sem_ges_faltante[['GP_GESTAN', 'sem_ges']])

Gestantes con sem_ges faltante después de la imputación:
Empty DataFrame
Columns: [GP_GESTAN, sem_ges]
Index: []


In [31]:
# Verificar valores únicos en GP_GESTAN
print("Valores únicos en GP_GESTAN:")
print(Data['GP_GESTAN'].unique())

Valores únicos en GP_GESTAN:
['2' '1']


In [32]:
# Verificar el tipo de dato de sem_ges
print("Tipo de dato de sem_ges:")
print(Data['sem_ges'].dtype)

# Verificar valores faltantes en sem_ges
print("Valores faltantes en sem_ges:")
print(Data['sem_ges'].isnull().sum())

Tipo de dato de sem_ges:
Int64
Valores faltantes en sem_ges:
989062


In [33]:
# Imputar 0 en sem_ges para no gestantes (GP_GESTAN == '2')
Data.loc[Data['GP_GESTAN'] == '2', 'sem_ges'] = Data.loc[Data['GP_GESTAN'] == '2', 'sem_ges'].fillna(0)

# Verificar valores faltantes en sem_ges después de la imputación
print("Valores faltantes en sem_ges después de la imputación:")
print(Data['sem_ges'].isnull().sum())

Valores faltantes en sem_ges después de la imputación:
27060


In [34]:
# Verificar no gestantes con sem_ges faltante después de la imputación
no_gestantes_con_sem_ges_faltante = Data[(Data['GP_GESTAN'] == '2') & (Data['sem_ges'].isna())]
print("No gestantes con sem_ges faltante después de la imputación:")
print(no_gestantes_con_sem_ges_faltante[['GP_GESTAN', 'sem_ges']])

No gestantes con sem_ges faltante después de la imputación:
Empty DataFrame
Columns: [GP_GESTAN, sem_ges]
Index: []


In [35]:
# Verificar gestantes con sem_ges faltante después de la imputación
gestantes_con_sem_ges_faltante = Data[(Data['GP_GESTAN'] == '1') & (Data['sem_ges'].isna())]
print("Gestantes con sem_ges faltante después de la imputación:")
print(gestantes_con_sem_ges_faltante[['GP_GESTAN', 'sem_ges']])

Gestantes con sem_ges faltante después de la imputación:
        GP_GESTAN  sem_ges
4               1     <NA>
29              1     <NA>
33              1     <NA>
34              1     <NA>
58              1     <NA>
...           ...      ...
1014279         1     <NA>
1014283         1     <NA>
1014302         1     <NA>
1014324         1     <NA>
1014351         1     <NA>

[27060 rows x 2 columns]


In [36]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
FEC_HOS,84.631325
COD_PAIS_R,64.490154
Pais_residencia,64.490154
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697
estrato,41.996697
fuente,41.996697
INI_SIN,7.284978
sem_ges,2.667652


In [37]:
import plotly.express as px

# Creación del gráfico de barras interactivo
fig = px.bar(
    MissingPercentage,
    x=MissingPercentage.index,
    y=MissingPercentage.values,
    labels={'x': 'Variables', 'y': 'Porcentaje de Valores Faltantes'},
    title='Porcentaje de Valores Faltantes por Variable',
    color=MissingPercentage.values,
    color_continuous_scale='Blues',
    text=MissingPercentage.values.round(2)
)

# Personalización del gráfico
fig.update_layout(
    xaxis_title='Variables',
    yaxis_title='Porcentaje de Valores Faltantes',
    title_font_size=24,
    xaxis_tickangle=-45,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(
        family="Georgia",
        size=14,
        color="black"
    ),
    title_font=dict(
        family="Georgia",
        size=24,
        color="black"
    ),
    coloraxis_colorbar=dict(
        title='Porcentaje',
        titlefont=dict(family="Georgia", size=14),
        tickfont=dict(family="Georgia", size=12)
    )
)

fig.update_traces(textposition='outside')

# Mostrar el gráfico
fig.show()

In [38]:
# Verificar el tipo de dato de FEC_HOS
print("Tipo de dato de FEC_HOS:")
print(Data['FEC_HOS'].dtype)

# Verificar valores faltantes en FEC_HOS
print("Valores faltantes en FEC_HOS:")
print(Data['FEC_HOS'].isnull().sum())

Tipo de dato de FEC_HOS:
datetime64[ns]
Valores faltantes en FEC_HOS:
858479


In [39]:
# Imputar "2099-12-31" en FEC_HOS para no hospitalizados (PAC_HOS == '2')
fecha_no_hospitalizado = pd.to_datetime('2099-12-31')
Data.loc[Data['PAC_HOS'] == '2', 'FEC_HOS'] = Data.loc[Data['PAC_HOS'] == '2', 'FEC_HOS'].fillna(fecha_no_hospitalizado)

# Verificar valores faltantes en FEC_HOS después de la imputación
print("Valores faltantes en FEC_HOS después de la imputación:")
print(Data['FEC_HOS'].isnull().sum())

Valores faltantes en FEC_HOS después de la imputación:
2


In [40]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
COD_PAIS_R,64.490154
Pais_residencia,64.490154
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697
estrato,41.996697
fuente,41.996697
INI_SIN,7.284978
sem_ges,2.667652
FECHA_NTO,2.251140


In [41]:
# Verificar valores únicos en PAC_HOS
print("Valores únicos en PAC_HOS:")
print(Data['PAC_HOS'].unique())

Valores únicos en PAC_HOS:
['2' '1']


In [42]:
import plotly.express as px

# Creación del gráfico de barras interactivo
fig = px.bar(
    MissingPercentage,
    x=MissingPercentage.index,
    y=MissingPercentage.values,
    labels={'x': 'Variables', 'y': 'Porcentaje de Valores Faltantes'},
    title='Porcentaje de Valores Faltantes por Variable',
    color=MissingPercentage.values,
    color_continuous_scale='Blues',
    text=MissingPercentage.values.round(2)
)

# Personalización del gráfico
fig.update_layout(
    xaxis_title='Variables',
    yaxis_title='Porcentaje de Valores Faltantes',
    title_font_size=24,
    xaxis_tickangle=-45,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(
        family="Georgia",
        size=14,
        color="black"
    ),
    title_font=dict(
        family="Georgia",
        size=24,
        color="black"
    ),
    coloraxis_colorbar=dict(
        title='Porcentaje',
        titlefont=dict(family="Georgia", size=14),
        tickfont=dict(family="Georgia", size=12)
    )
)

fig.update_traces(textposition='outside')

# Mostrar el gráfico
fig.show()

In [43]:
# Verificar valores únicos en Departamento_residencia
print("Valores únicos en Departamento_residencia:")
print(Data['Departamento_residencia'].unique())

Valores únicos en Departamento_residencia:
['VALLE' 'NARIÑO' 'BOGOTA' 'SANTANDER' 'ANTIOQUIA' 'BOLIVAR' 'HUILA'
 'CAQUETA' 'BOYACA' 'CAUCA' 'META' 'GUAJIRA' 'CESAR' 'CUNDINAMARCA'
 'TOLIMA' 'ATLANTICO' 'CALDAS' 'QUINDIO' 'GUAVIARE' 'SUCRE' 'MAGDALENA'
 'PUTUMAYO' 'RISARALDA' 'CORDOBA' 'VICHADA' 'NORTE SANTANDER' 'CASANARE'
 'ARAUCA' 'AMAZONAS' 'GUAINIA' 'CHOCO' 'EXTERIOR' 'SAN ANDRES'
 'PROCEDENCIA DESCONOCIDA' 'VAUPES']


In [44]:
# Lista de departamentos válidos de Colombia
departamentos_colombia = [
    'VALLE', 'MARINO', 'BOGOTA', 'SANTANDER', 'ANTIOQUIA', 'BOLIVAR', 'HUILA',
    'CAQUETA', 'BOYACA', 'CAUCA', 'META', 'GUAJIRA', 'CESAR', 'CUNDINAMARCA',
    'TOLIMA', 'ATLANTICO', 'CALDAS', 'QUINDIO', 'GUAVIARE', 'SUCRE', 'MAGDALEMA',
    'PUTUNAVO', 'RISARAIDA', 'CORDOBA', 'VICHADA', 'MORTE SANTANDER', 'CASAWARE',
    'ARAUCA', 'AMAZONAS', 'GUATNIA', 'CHOCO', 'SAN ANDRES', 'VAUPES'
]

In [45]:
# Código y nombre de Colombia (como cadenas)
codigo_colombia = "170"  # Asegúrate de que sea una cadena
pais_colombia = "COLOMBIA"

# Condición: Departamento_residencia es un departamento de Colombia y COD_PAIS_R/Pais_residencia son nulos
condicion = (Data['Departamento_residencia'].isin(departamentos_colombia)) & (Data['COD_PAIS_R'].isna() | Data['Pais_residencia'].isna())

# Imputar
Data.loc[condicion, 'COD_PAIS_R'] = codigo_colombia
Data.loc[condicion, 'Pais_residencia'] = pais_colombia

In [46]:
# Condición: Departamento_residencia es EXTERIOR o PROCEDENCIA DESCONOCIDA
condicion_exterior_desconocido = Data['Departamento_residencia'].isin(['EXTERIOR', 'PROCEDENCIA DESCONOCIDA'])

# Dejar los valores faltantes en COD_PAIS_R y Pais_residencia
Data.loc[condicion_exterior_desconocido, 'COD_PAIS_R'] = pd.NA
Data.loc[condicion_exterior_desconocido, 'Pais_residencia'] = pd.NA

In [217]:
# Condición: Departamento_residencia no es un departamento de Colombia y no es EXTERIOR o PROCEDENCIA DESCONOCIDA
condicion_faltante_o_no_valido = (
    ~Data['Departamento_residencia'].isin(departamentos_colombia) &
    Data['Departamento_residencia'].notna() &
    ~Data['Departamento_residencia'].isin(['EXTERIOR', 'PROCEDENCIA DESCONOCIDA'])
)

# Dejar los valores faltantes en COD_PAIS_R y Pais_residencia
Data.loc[condicion_faltante_o_no_valido, 'COD_PAIS_R'] = pd.NA
Data.loc[condicion_faltante_o_no_valido, 'Pais_residencia'] = pd.NA

In [47]:
print("Valores faltantes en COD_PAIS_R después de la imputación:")
print(Data['COD_PAIS_R'].isnull().sum())

print("Valores faltantes en Pais_residencia después de la imputación:")
print(Data['Pais_residencia'].isnull().sum())

Valores faltantes en COD_PAIS_R después de la imputación:
86917
Valores faltantes en Pais_residencia después de la imputación:
86917


In [48]:
import pandas as pd

# Calcular el porcentaje de valores faltantes por columna
MissingPercentage = (Data.isnull().sum() / len(Data)) * 100
MissingPercentage = MissingPercentage[MissingPercentage > 0]  # Solo columnas con datos faltantes
MissingPercentage = MissingPercentage.sort_values(ascending=False)  # Ordenar de mayor a menor

display((MissingPercentage))

,0
nacionalidad,41.996697
nombre_nacionalidad,41.996697
nom_grupo,41.996697
estrato,41.996697
fuente,41.996697
COD_PAIS_R,8.568527
Pais_residencia,8.568527
INI_SIN,7.284978
sem_ges,2.667652
FECHA_NTO,2.251140


In [49]:
import plotly.express as px

# Creación del gráfico de barras interactivo
fig = px.bar(
    MissingPercentage,
    x=MissingPercentage.index,
    y=MissingPercentage.values,
    labels={'x': 'Variables', 'y': 'Porcentaje de Valores Faltantes'},
    title='Porcentaje de Valores Faltantes por Variable',
    color=MissingPercentage.values,
    color_continuous_scale='Blues',
    text=MissingPercentage.values.round(2)
)

# Personalización del gráfico
fig.update_layout(
    xaxis_title='Variables',
    yaxis_title='Porcentaje de Valores Faltantes',
    title_font_size=24,
    xaxis_tickangle=-45,
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(
        family="Georgia",
        size=14,
        color="black"
    ),
    title_font=dict(
        family="Georgia",
        size=24,
        color="black"
    ),
    coloraxis_colorbar=dict(
        title='Porcentaje',
        titlefont=dict(family="Georgia", size=14),
        tickfont=dict(family="Georgia", size=12)
    )
)

fig.update_traces(textposition='outside')

# Mostrar el gráfico
fig.show()

In [50]:
# Verificar si los valores faltantes están en los mismos registros
mismas_filas_faltantes = Data[['nacionalidad', 'nombre_nacionalidad', 'nom_grupo', 'estrato', 'fuente']].isnull().all(axis=1)
print("Número de registros con valores faltantes en todas las variables:")
print(mismas_filas_faltantes.sum())

Número de registros con valores faltantes en todas las variables:
426004


In [51]:
# Verificar valores faltantes en FEC_HOS
print("Valores faltantes en nacionalidad:")
print(Data['nacionalidad'].isnull().sum())

print("Valores faltantes en nombre_nacionalidad:")
print(Data['nombre_nacionalidad'].isnull().sum())

print("Valores faltantes en nom_grupo:")
print(Data['nom_grupo'].isnull().sum())

print("Valores faltantes en estrato:")
print(Data['estrato'].isnull().sum())

print("Valores faltantes en fuente:")
print(Data['fuente'].isnull().sum())


Valores faltantes en nacionalidad:
426004
Valores faltantes en nombre_nacionalidad:
426004
Valores faltantes en nom_grupo:
426004
Valores faltantes en estrato:
426004
Valores faltantes en fuente:
426004


In [52]:
# Verificar si los valores faltantes están concentrados en un período específico
Data['FEC_NOT'] = pd.to_datetime(Data['FEC_NOT'], errors='coerce')  # Convertir a datetime si no lo está
Data['ANO'] = Data['FEC_NOT'].dt.year  # Extraer el año

# Verificar la distribución de valores faltantes por año
print("Distribución de valores faltantes por año:")
print(Data.groupby('ANO')[['nacionalidad', 'nombre_nacionalidad', 'nom_grupo', 'estrato', 'fuente']].apply(lambda x: x.isnull().mean()))

Distribución de valores faltantes por año:
      nacionalidad  nombre_nacionalidad  nom_grupo   estrato    fuente
ANO                                                                   
2012      1.000000             1.000000   1.000000  1.000000  1.000000
2013      1.000000             1.000000   1.000000  1.000000  1.000000
2014      1.000000             1.000000   1.000000  1.000000  1.000000
2015      1.000000             1.000000   1.000000  1.000000  1.000000
2016      1.000000             1.000000   1.000000  1.000000  1.000000
2017      0.999352             0.999352   0.999352  0.999352  0.999352
2018      0.017778             0.017778   0.017778  0.017778  0.017778
2019      0.000000             0.000000   0.000000  0.000000  0.000000
2020      0.000000             0.000000   0.000000  0.000000  0.000000
2021      0.000000             0.000000   0.000000  0.000000  0.000000
2022      0.000000             0.000000   0.000000  0.000000  0.000000


In [53]:
# Verificar la distribución de valores faltantes por año
print("Distribución de valores faltantes por año:")
print(Data.groupby('ANO')[['nacionalidad', 'nombre_nacionalidad', 'nom_grupo', 'estrato', 'fuente']].apply(lambda x: x.isnull().mean()))

Distribución de valores faltantes por año:
      nacionalidad  nombre_nacionalidad  nom_grupo   estrato    fuente
ANO                                                                   
2012      1.000000             1.000000   1.000000  1.000000  1.000000
2013      1.000000             1.000000   1.000000  1.000000  1.000000
2014      1.000000             1.000000   1.000000  1.000000  1.000000
2015      1.000000             1.000000   1.000000  1.000000  1.000000
2016      1.000000             1.000000   1.000000  1.000000  1.000000
2017      0.999352             0.999352   0.999352  0.999352  0.999352
2018      0.017778             0.017778   0.017778  0.017778  0.017778
2019      0.000000             0.000000   0.000000  0.000000  0.000000
2020      0.000000             0.000000   0.000000  0.000000  0.000000
2021      0.000000             0.000000   0.000000  0.000000  0.000000
2022      0.000000             0.000000   0.000000  0.000000  0.000000


**Valores faltantes por año:**

- 2012-2017: El 100% de los registros tienen valores faltantes en estas variables.

- 2017: El 99.94% de los registros tienen valores faltantes.

- 2018: Solo el 1.78% de los registros tienen valores faltantes.

- 2019-2022: No hay valores faltantes.

Registros con valores faltantes: Hay 426,004 registros con valores faltantes en todas las variables.

### Justificación para mantener el **estrato**:
El estrato socioeconómico es una variable fundamental en estudios que buscan analizar desigualdades o segmentar poblaciones según su nivel de ingresos y acceso a servicios. Su relevancia radica en su capacidad para capturar diferencias estructurales en la distribución de recursos y oportunidades, lo que la convierte en un predictor clave en modelos socioeconómicos (CEPAL, 2018). Aunque existan datos faltantes, la imputación de valores es preferible a su eliminación, ya que permite preservar la información crítica que esta variable aporta al análisis, evitando así sesgos y pérdida de información valiosa (Little & Rubin, 2019).

### Justificación para mantener la **nacionalidad**:
La nacionalidad es una variable esencial en contextos donde la diversidad cultural, migratoria o legal es relevante, ya que permite identificar patrones demográficos y de integración social. Su inclusión es particularmente importante en estudios de políticas públicas o análisis de poblaciones heterogéneas (OIM, 2020). A pesar de la presencia de datos faltantes, técnicas de imputación categórica, como la moda o modelos predictivos, permiten estimar valores faltantes de manera efectiva, manteniendo la integridad del análisis y evitando la pérdida de información clave (Van Buuren, 2018).

### Justificación para eliminar otras variables:
La eliminación de otras tres variables se justifica cuando estas no aportan un valor analítico significativo o presentan una alta tasa de datos faltantes, lo que dificulta su imputación confiable. Reducir la dimensionalidad del conjunto de datos mejora la eficiencia del modelo y minimiza el riesgo de sobreajuste, especialmente cuando las variables eliminadas no están teóricamente relacionadas con el objetivo del estudio (Hair et al., 2019). Esta práctica es común en análisis estadísticos para mantener solo las variables más relevantes y robustas.

---

### Fuentes:
- CEPAL. (2018). *La ineficiencia de la desigualdad*. Disponible en: [https://www.cepal.org](https://www.cepal.org)
- Little, R. J. A., & Rubin, D. B. (2019). *Statistical Analysis with Missing Data*. Wiley.
- OIM. (2020). *Informe sobre las migraciones en el mundo*. Disponible en: [https://www.iom.int](https://www.iom.int)
- Van Buuren, S. (2018). *Flexible Imputation of Missing Data*. CRC Press.
- Hair, J. F., Black, W. C., Babin, B. J., & Anderson, R. E. (2019). *Multivariate Data Analysis*. Pearson.



In [124]:
import pandas as pd

# Supongamos que tienes un DataFrame llamado df
# Eliminar las columnas no deseadas
Data = Data.drop(columns=['nombre_nacionalidad', 'nom_grupo', 'fuente'])

# Verificar el resultado
print(Data.head())

  CONSECUTIVE    FEC_NOT  SEMANA   ANO  EDAD UNI_MED nacionalidad SEXO  \
0     2934558 2014-06-03      22  2014    15       1         <NA>    F   
1     2934568 2014-03-24      13  2014    17       1         <NA>    F   
2     2934589 2014-08-12      33  2014    17       1         <NA>    F   
3     2934609 2014-05-18      21  2014    17       1         <NA>    M   
4     2934620 2014-03-26       5  2014    17       1         <NA>    F   

  COD_PAIS_O COD_DPTO_O  ... nom_est_f_caso  \
0        170         76  ...     Sospechoso   
1        170         52  ...     Sospechoso   
2        170         11  ...     Sospechoso   
3        170         68  ...     Sospechoso   
4        170         68  ...     Sospechoso   

                                            Nom_upgd Pais_ocurrencia  \
0                           CLINICA SALUDCOOP NOVENA        COLOMBIA   
1                        CENTRO DE SALUD SAN VICENTE        COLOMBIA   
2               ESE HOSPITAL SIMON BOLIVAR III NIVEL    

In [125]:
print(f"Cantidad de registros: {Data.shape[0]}, Cantidad de variables: {Data.shape[1]}")

Cantidad de registros: 1014375, Cantidad de variables: 54


## Iterative inputer


In [141]:
import pandas as pd
import numpy as np

# Convertir la columna a string (si no lo es ya)
Data['estrato'] = Data['estrato'].astype(str)

# Eliminar espacios adicionales y reemplazar valores no válidos con NaN
Data['estrato'] = Data['estrato'].str.strip().replace(['', 'nan', 'NA', '<NA>', 'NaN'], np.nan)

# Convertir a numérico
Data['estrato'] = Data['estrato'].astype(float)